### DISCUSSION TITLE
iperf commands - for throughput / traffic

### DISCUSSION THREAD

**Discussion by jchellap** [openthread/openthread#9507] [General]
Hi
With the open thread source code, can we use iperf tool and run througput or kind of tests?
Any suggestions for the same?

**Comment by jwhui** [comment]
OpenThread itself does not implement an `iperf` endpoint. However, given that `iperf` operates on top of IP, OpenThread can transport `iperf` message. For example, you could run `iperf` between to Linux hosts connected via Thread.

Alternatively, OpenThread does provide some related performance utilities - see [`tcp`](https://github.com/openthread/openthread/blob/main/src/cli/README_TCP.md) and [`ping`](https://github.com/openthread/openthread/blob/main/src/cli/README.md#ping-async--i-source-ipaddr-size-count-interval-hoplimit-timeout) CLI commands.

**Reply by jchellap** [reply]
@jwhui , Thanks! we are looking tools for throughput. However, given that iperf operates on top of IP, OpenThread can transport iperf message.
Does this mean I can use iperf command from the iperf tool or like CLI command should be used? Please quote example links which can be used to further understand on this. Thanks for the support

**Comment by Irving-cl** [comment]
@jchellap 

I actually did this before. However it was not operated on universally availble devices like raspberry pi. But yes, it can run on your device (linux-based devices). I can show you the steps to run it:
* You need to first prepare 2 Thread devices (must be host+RCP mode). Because you need to run iperf on the linux system. Things like `ot-cli-ftd`, `ot-ncp-ftd` cannot run iperf.
* Setup the Thread network on your thread devices, make sure they are connected. You can ping to ensure this. Then you will see the `wpan` interface in the system of the two devices.
* You may need to turn off the firewall on your device. This depends on your device. For example, run this on both of your devices: `ip6tables -I INPUT -p udp --dport 5001 -j ACCEPT`
* Get the Thread IPv6 address on your receive end. I usually use the mleid address. You can get it ot-ctl: `ot-ctl ipaddr mleid`.
* Start the iperf server on your receiving end: `iperf -V -s -u -i 1 -U`. You may adjust the options based on your needs.
* Start the iperf client on your sending end: `iperf -V -c {mleid} -t {time_to_run_in_seconds} -u -i 1 -b {bandwidth_kbps} -l {udp_packet_length}`

This should work. But since there may be many differences between your devices and mine including the iperf version, you may need some debugging.

So for your question:
> Does this mean I can use iperf command from the iperf tool or like CLI command should be used? 

You should use iperf tool.

**Reply by jchellap** [reply]
That helps a lot. Thanks much for the guidance.
I was able to get tcp packets working. UDP failed for some reasons with the above mleid

**Reply by Irving-cl** [reply]
> UDP failed for some reasons with the above mleid

What's the output of `iperf` of both sides?

**Reply by jchellap** [reply]
using 2ce thread branch...
On server side command : iperf -s -t -p 5123 -i 1 -l 65500 -V
------------------------------------------------------------
Server listening on UDP port 5123
Receiving 65500 byte datagrams
UDP buffer size:  208 KByte (default)
------------------------------------------------------------
[  3] local fd00:db8::fae4:3bff:feaf:c496 port 5123 connected with fd00:db8::fae4:3bff:feaf:c447 port 44159
[ ID] Interval       Transfer     Bandwidth        Jitter   Lost/Total Datagrams
[  3]  0.0- 1.0 sec  0.00 Bytes  0.00 bits/sec   0.000 ms    0/    0 (0%)
[  3]  1.0- 2.0 sec  0.00 Bytes  0.00 bits/sec   0.000 ms    0/    0 (0%)
[  3]  2.0- 3.0 sec  0.00 Bytes  0.00 bits/sec   0.000 ms    0/    0 (0%)
[  3]  3.0- 4.0 sec  0.00 Bytes  0.00 bits/sec   0.000 ms    0/    0 (0%)
[  3]  4.0- 5.0 sec  0.00 Bytes  0.00 bits/sec   0.000 ms    0/    0 (0%)
[  3]  5.0- 6.0 sec  0.00 Bytes  0.00 bits/sec   0.000 ms    0/    0 (0%)
[  3]  6.0- 7.0 sec  0.00 Bytes  0.00 bits/sec   0.000 ms    0/    0 (0%)
[  3]  7.0- 8.0 sec  0.00 Bytes  0.00 bits/sec   0.000 ms    0/    0 (0%)
[  3]  8.0- 9.0 sec  0.00 Bytes  0.00 bits/sec   0.000 ms    0/    0 (0%)
[  3]  9.0-10.0 sec  0.00 Bytes  0.00 bits/sec   0.000 ms    0/    0 (0%)
[  3]  0.0-10.1 sec  0.00 Bytes  0.00 bits/sec   0.000 ms    0/    0 (0%)
[  4] local fd00:db8::fae4:3bff:feaf:c496 port 5123 connected with fd00:db8::fae4:3bff:feaf:c447 port 44159
[  4]  0.0- 1.0 sec  0.00 Bytes  0.00 bits/sec   0.000 ms    0/    0 (0%)
[  4]  1.0- 2.0 sec  0.00 Bytes  0.00 bits/sec   0.000 ms    0/    0 (0%)
[  4]  2.0- 3.0 sec  0.00 Bytes  0.00 bits/sec   0.000 ms    0/    0 (0%)
[  4]  3.0- 4.0 sec  0.00 Bytes  0.00 bits/sec   0.000 ms    0/    0 (0%)
[  4]  4.0- 5.0 sec  0.00 Bytes  0.00 bits/sec   0.000 ms    0/    0 (0%)
[  4]  5.0- 6.0 sec  0.00 Bytes  0.00 bits/sec   0.000 ms    0/    0 (0%)
[  4]  6.0- 7.0 sec  0.00 Bytes  0.00 bits/sec   0.000 ms    0/    0 (0%)
[  4]  7.0- 8.0 sec  0.00 Bytes  0.00 bits/sec   0.000 ms    0/    0 (0%)
[  4]  8.0- 9.0 sec  0.00 Bytes  0.00 bits/sec   0.000 ms    0/    0 (0%)
[  4]  9.0-10.0 sec  0.00 Bytes  0.00 bits/sec   0.000 ms    0/    0 (0%)
[  4]  0.0-10.1 sec  0.00 Bytes  0.00 bits/sec   0.000 ms    0/    0 (0%)
read failed: Connection refused

On Client side:
iperf -c fd00:db9:0:0:7f03:4bb8:6f37:2ee0 -V -u -p 5123 -i 1 -t 20 -b 1K
WARNING: delay too large, reducing from 11.8 to 1.0 seconds.
------------------------------------------------------------
Client connecting to fd00:db9:0:0:7f03:4bb8:6f37:2ee0, UDP port 5123
Sending 1470 byte datagrams, IPG target: 11760000.00 us (kalman adjust)
UDP buffer size:  208 KByte (default)
------------------------------------------------------------
[  3] local fd00:db8::fae4:3bff:feaf:c447 port 44159 connected with fd00:db9::7f03:4bb8:6f37:2ee0 port 5123
[ ID] Interval       Transfer     Bandwidth
[  3]  0.0- 1.0 sec  2.87 KBytes  23.5 Kbits/sec
[  3]  1.0- 2.0 sec  1.44 KBytes  11.8 Kbits/sec
[  3]  2.0- 3.0 sec  1.44 KBytes  11.8 Kbits/sec
[  3]  3.0- 4.0 sec  1.44 KBytes  11.8 Kbits/sec
[  3]  4.0- 5.0 sec  1.44 KBytes  11.8 Kbits/sec
[  3]  5.0- 6.0 sec  1.44 KBytes  11.8 Kbits/sec
[  3]  6.0- 7.0 sec  1.44 KBytes  11.8 Kbits/sec
[  3]  7.0- 8.0 sec  1.44 KBytes  11.8 Kbits/sec
[  3]  8.0- 9.0 sec  1.44 KBytes  11.8 Kbits/sec
[  3]  9.0-10.0 sec  1.44 KBytes  11.8 Kbits/sec
[  3] 10.0-11.0 sec  1.44 KBytes  11.8 Kbits/sec
[  3] 11.0-12.0 sec  1.44 KBytes  11.8 Kbits/sec
[  3] 12.0-13.0 sec  1.44 KBytes  11.8 Kbits/sec
[  3] 13.0-14.0 sec  1.44 KBytes  11.8 Kbits/sec
[  3] 14.0-15.0 sec  1.44 KBytes  11.8 Kbits/sec
[  3] 15.0-16.0 sec  1.44 KBytes  11.8 Kbits/sec
[  3] 16.0-17.0 sec  1.44 KBytes  11.8 Kbits/sec
[  3] 17.0-18.0 sec  1.44 KBytes  11.8 Kbits/sec
[  3] 18.0-19.0 sec  1.44 KBytes  11.8 Kbits/sec
[  3] 19.0-20.0 sec  1.44 KBytes  11.8 Kbits/sec
[  3]  0.0-21.0 sec  30.1 KBytes  11.8 Kbits/sec
[  3] Sent 21 datagrams
[  3] WARNING: did not receive ack of last datagram after 10 tries.

**Reply by Irving-cl** [reply]
Have you disabled firewall for 5123 port? In my example, it's 5001. You need to change.

Another thing is, in your output
> [ 3] local fd00:db8::fae4:3bff:feaf:c496 port 5123 connected with fd00:db8::fae4:3bff:feaf:c447 port 44159

The address on server `fd00:db8::fae4:3bff:feaf:c496` is different from the one you used: `fd00:db9:0:0:7f03:4bb8:6f37:2ee0`. I met this issue before (though I don't know why). The solution is bind the address on server side by `-I`(Capital i) option: `-I fd00:db9:0:0:7f03:4bb8:6f37:2ee0`

BTW, the maximum IP packet length for Thread is 1280. If larger than that, I think it will be split into multiple IP packets. So when setting the udp packet length, you may want to use 1232 here to ensure the whole udp packet is encapsulated in one IP packet to achieve higher throughput.

**Reply by jchellap** [reply]
For port 5123, i checked for firewall with the command you gave, no response. So I assume disabled properly.
fd00:db9:0:0:7f03:4bb8:6f37:2ee0 is mleid of server. You mean do a udp bind before issuing the iperf commands for UDP?
Didnt quiet follow this -->  The solution is bind the address on server side by -I(Capital i) option: -I fd00:db9:0:0:7f03:4bb8:6f37:2ee0

Also, the max packet length how can i ensure its within the limit?
One more query, the above ones I tested with Router as a child device. What about SED and SSED since they will be sleeping device, how will the parameters differ? Any example here will help understand.

**Reply by Irving-cl** [reply]
> Didnt quiet follow this --> The solution is bind the address on server side by -I(Capital i) option: -I fd00:db9:0:0:7f03:4bb8:6f37:2ee0

I mean adding the option in the command when starting iperf server: `iperf -s -t -p 5123 -i 1 -l 1232 -V -I fd00:db9:0:0:7f03:4bb8:6f37:2ee0`

> Also, the max packet length how can i ensure its within the limit?

You need to sniffer the packets. Or you can check the log from thread stack on your device. `1232` is the maximum value with my experience. If you want to ensure it's within one IP packet without other effort, I think you can simply use 1200.

**Reply by Irving-cl** [reply]
> What about SED and SSED since they will be sleeping device, how will the parameters differ? Any example here will help understand.

Nothing needs to be changed. When using UDP, if you specified the bandwidth on client, it will try the best to send to reach that value. With SED or SSED, you may expect lower throughput result.

**Reply by jchellap** [reply]
Hi,
I was able to get it working for UDP also. I was using iperf. After using iperf3 and use -B to bind the address, I am able to get it working for both UDP and TCP. Thanks for the immediate guidance.

**Reply by jchellap** [reply]
One other question is what is the typical throughput value we can get for UDP using Open thread RCP design?

**Reply by Irving-cl** [reply]
> One other question is what is the typical throughput value we can get for UDP using Open thread RCP design?

From what I got, you can usually get 60~70 Kbps for UDP throughput. We also achieved 80+(near 90) Kbps on some devices. Note that the performance also depends on the environment. You may want to use a RF-shielded space to run the test to achieve high throughput

**Reply by jchellap** [reply]
Thanks much for some reference data. This helps. I hope the measurement is based on the Server side for the RX right?

**Reply by jchellap** [reply]
What should be the bitrate at the client side? i used bit rate as 1K
iperf3 -c IPV6 address -V -u -p 5123 -i 1 -t 20 -b 1K -l 1200

**Reply by Irving-cl** [reply]
`-b` is the highest bandwidth you want. Iperf won't exceed the number. Also it may not reach the number. If you set 1K, it's 1Kbps at most. You can use `80K` here. If it can reach 80K, you can try a larger value.

**Comment by jchellap** [comment]
Further one query, When i use Leader as server and Router as child and client, i see UDP messages going through between client and server...with some throughput values.
However when i start iperf3 on child as server and Leader as client, it doesnt work. No data exchange.
Is it not to be used? Any explanation for not working will help me understand the right method of usage.

**Reply by zhanglongxia** [reply]
I also met iperf3 issues in recent days. Here are the methods that I resolved issues. Hope it could help you.

1.  IPerf3 will create a TCP connection before transmitting UDP messages. Thread stack has enabled the TCP feature by default. It causes iperf3 to be unable to successfully create a TCP connection. So we can disable the [OPENTHREAD_CONFIG_TCP_ENABLE](https://github.com/openthread/openthread/blob/main/src/core/config/ip6.h#L172) in Thread stack.
2. The SRP server will add the anycast address to the Thread network interface. The Linux network stack has a chance to select the anycast address as the source address of the Thread IP packet. Try to run the command `ot-ctl srp server disable` to resolve this issue.
3. On the iperf3 server side, bind a Thread network IP address. For example, iperf3 server:  `iperf3 -s -i 1 -B fd08:3400:dce1:8a22:5645:80f8:66a7:27f1`, iperf3 client: `iperf3 -6 -c fd08:3400:dce1:8a22:5645:80f8:66a7:27f1 -u -b 60k -i 1 -t 30`.
4. Check the firewall on your devices to make sure that Thread IP packets are not blocked by the firewall.
5. Use the command `ping6` on both devices to ping each other to make sure that the routing table is correct.
6. Use `tcpdump -i wpan0` to monitor the traffic on the `wpan0` interface to check whether the Thread IP traffic is as expected.

**Reply by jchellap** [reply]
Thanks will try the same.
I am able to successfully test when Leader is server and Child as Client.
I see no packet transmission/transaction when Leader is client and Child as Server. Is this usecase correct at all?

**Reply by zhanglongxia** [reply]
Iperf3 should work regardless of whether the Iperf3 server is leader or child. You can check the OpenThread log and the output of `tcpdump -i wpan0` on both leader and child to find where the IP packet is dropped. That will help you find out the root cause.

**Comment by jchellap** [comment]
I tried disabling the [OPENTHREAD_CONFIG_TCP_ENABLE](https://github.com/openthread/openthread/blob/main/src/core/config/ip6.h#L172) in Thread and also ot-ctl srp server disable.
When i trigger iperf as client on Leader, i see some packets that is decrypted as TCP retransmission.

![image](https://github.com/openthread/openthread/assets/131742063/7e99c773-ecdb-4e9b-a6e6-ae8f504d4266)

Anyone came across such issue?

**Reply by zhanglongxia** [reply]
From the log, the TCP SYNC Request packet has been sent to the Child and the Child has replied the ACK. It means the Child has received the packet. So you need to check the OpenThread log and capture the IP packets on the `wpan0` interface on the Child device to find out where the TCP SYNC Request packet is dropped.

**Reply by jchellap** [reply]
When we send iperf3 to transmit packets as UDP, why is it transmitting TCP packets instead of UDP? This is observed only when the Leader is made as client. May be some more macros to be disabled? The packets transmitted by the Child device when started as iperf3 client is proper UDP packet. Hence the query

**Reply by zhanglongxia** [reply]
Iperf3 uses TCP to create a connection to start the testing. See https://github.com/esnet/iperf/issues/1331#issuecomment-1120186827

**Reply by jchellap** [reply]
With the openthread logs for the child node configured as server says as below, there was no message for tcpdump -i wpan0 for iperf server device 
I see otbr web service failing...
Nov  2 21:30:23 localhost systemd: otbr-web.service: Failed at step EXEC spawning /usr/sbin/otbr-web: Exec format error
Nov  2 21:30:23 localhost systemd: otbr-web.service: Main process exited, code=exited, status=203/EXEC
Nov  2 21:30:23 localhost systemd: otbr-web.service: Unit entered failed state.
Nov  2 21:30:23 localhost audit: SERVICE_STOP pid=1 uid=0 auid=4294967295 ses=4294967295 msg='unit=otbr-web comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=failed'
Nov  2 21:30:23 localhost systemd: otbr-web.service: Failed with result 'exit-code'.
Nov  2 21:30:23 localhost otbr-agent[2945]: 00:48:05.719 [I] MeshForwarder-: Received IPv6 UDP msg, len:91, chksum:ee9a, ecn:no, from:166e0a0000000001, sec:no, prio:net, rss:-92.0
Nov  2 21:30:23 localhost otbr-agent[2945]: 00:48:05.719 [I] MeshForwarder-:     src:[fe80:0:0:0:146e:a00:0:1]:19788
Nov  2 21:30:23 localhost otbr-agent[2945]: 00:48:05.719 [I] MeshForwarder-:     dst:[ff02:0:0:0:0:0:0:1]:19788
Nov  2 21:30:23 localhost otbr-agent[2945]: 00:48:05.719 [I] Mle-----------: Receive Advertisement (fe80:0:0:0:146e:a00:0:1,0x0400)
Nov  2 21:30:28 localhost systemd: otbr-web.service: Service hold-off time over, scheduling restart.
Nov  2 21:30:28 localhost audit: SERVICE_START pid=1 uid=0 auid=4294967295 ses=4294967295 msg='unit=otbr-web comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'

**Reply by jchellap** [reply]
Hi, further i have one query please guide.
When i ping the devices each other within ot-ctl. I see leader and child node works fine.
However, when i ping from terminal(not within ot-ctl) ping shows for child node
Packets are dropped by the child node and the message log shows as below:

![image](https://github.com/openthread/openthread/assets/131742063/bca04cc2-fa64-430b-8ce0-e74f0bfbb720)


Any feature level macro to be enabled or disabled? Also what does error:parse mean? where can i find references for this?

**Reply by jwhui** [reply]
Parse errors are often due to a mismatch in the mesh-local prefix configuration.

Can you verify that the Active Operational Dataset is exactly the same on your Thread devices?

**Reply by jchellap** [reply]
Data set is same in both the devices. Just a thought.. If active operational data set  is not matching then When child is made as Client and Leader as server also should not work right? or ping from Child device also should not work correct?
I see ping not working only from the Leader and being dropped by the child.

**Reply by jchellap** [reply]
Just an update. With 2ce tag for ot-br-posix and open thread source code, i see that Leader configured as Client for UDP data transmission is not working and child device as server.
However, thread-reference-20221027 git commit, the  child device as server is working as expected. Any particular commit that i can cherry-pick or should i move to the latest code base?

**Reply by jwhui** [reply]
Unfortunately, I'm not aware of any specific commit that would cause this issue. Would you be willing to do a git-bisect to pinpoint the commit?

**Reply by jchellap** [reply]
I will need sometime to check, the quick resolution I found was to move to later commit and continue. The other issues I found is crashing my tool due to extra literals used in the print in daemon.cpp and  net_if.cpp. 
In daemon.cpp
 otLogInfoPlat("Session socket is ready"); --> this had additional literal such %s. 
In net_if.cpp --> may be this is fixed already.. just for your info
otLogWarnPlat("[netif] Successfully added an external route %s in kernel:", prefixString); -> has extra %s

For the iperf.. will need sometime to find exactly which commit will caused the issue. Since there was some code changes in ip6.cpp interms of structure usage, i felt patching them will make it more hard to build the source code. The issue was on the receiving side of the iperf packet.

**Reply by jwhui** [reply]
The format string issues were fixed in #8353 and #8356.

**Reply by jchellap** [reply]
Thanks for sharing the above link. Had missed one formatting error in my copy.
In general how do we sync up such changes with the open source code? Can you recommend/suggest? Should we move to the latest tags etc? That is what I was following

**Comment by jchellap** [comment]
One other query, I saw your post in Linked about iphone hosting Open Thread.. In mobile phones, the device is going to be Border Router or What kind of roles it takes? Any link that can help me understand in real time use cases which type of nodes is selected?
For SSED kind of devices, which is the real time product that can be used?
